docs/design-tech/sync.md

```diff
@@ -1,89 +1,263 @@
 
-1. `calculate_work`, iterate through the results and hand each `work_unit` to execute work. Many execute-work errors are
-   retry-able.
-   For example, if a peer edited a file during your sync, you may fail to push your changes. This is fine, however,
-   because you'll pull those changes the next time this loop executes, merge, and then push your changes up (likely 
-   successfully this time).
-2. If there were no errors, `calculate_work` again. `execute_work` will likely cause there to be new updates. However,
-   we likely know about these updates because we caused them. We `calculate_work` to prevent erroneous steps next time 
-   we sync. 
-3. Repeat until `calculate_work` is empty.
+Pulling the metadata for a file and pulling the content for a file are, in terms of performance, categorically different. A file's contents are generally many orders of magnitude larger than file metadata. When we pull an update for a file, we need to avoid re-pulling the content for the file if it hasn't changed (e.g. if the file was only renamed or moved since the client last checked with the server). To this end, alongside `metadata_version`, we store a `content_version` in each metadata which is a server-assigned timestamp of the file's most recent content update. When a client pulls a metadata update, it also pulls the file content if and only if the `content_version` has changed since it last pulled.
 
-## `execute_work`
+Clients store two versions of the file tree, `base` and `local`. In order to save space, we use semantics similar to copy-on-write where the version of a file on `local` is not saved if it is identical to the version on `base`. Clients interpret the absence of a `local` version to mean that `local` and `base` are identical. Additionally, we compress files, both before saving to disk and before pushing to the server.
 
-### `server -> core`
+The metadata for deleted files must be eventually pruned on clients so that the amount of space taken on users' devices scales with the number of not-deleted files rather than the total number of files ever created. Deleted files cannot be immediately pruned because the client needs to remember to push the deletion to the server on the next sync. The server keeps track of all file metadata forever (though it deletes contents of deleted files) so that clients can sync after being offline indefinitely and still receive deletion events without the complexity of the server needing to know which clients exist and which updates they have each received.
```

Comment:
> cannot be immediately pruned because the client needs to remember to push the deletion to the server on the next sync.

however it's content can, correct?

delete is a permanent operation.

Okay so what about in the recursive sense, what if that file was rescued by someone else out of a folder that was deleted by you.

That document will need to be undeleted. Perhaps an area that requires further consideration.

Could be frusterating to delete a file because your phone is out of space, and then the space doesn't free up, and you're confused about why.

Reply:
>  Note that deleted documents can have their contents pruned immediately - it is only important to remember the metadata until the deletion is pushed.

okay I see this is covered later, let's see if the folder case is covered

Reply:
okay cool, I'll just resolve this one lol